### PR TITLE
encryption: add support for encryption to `writeFileSyncing`.

### PR DESCRIPTION
**Pull request by ghost**
`writeFileSyncing()` now will be able to write encrypted content to RocksDB's
env. This commit is part of `use encryption for all local disk usage (non-logs)`.

Issue: #19783.
Release note: None

### CODE REVIEW

**Comment by benesch** [comment]
Woohoo, this is looking really good. The one piece of feedback so far—I haven't reviewed in detail yet—is to extract the AppendFile and CloseFile methods into a separate object. Roughly:

```
type Engine interface {
  OpenFile() (DBFile, error)
}

type DBFile interface {
  AppendFile(contents []byte) error
  CloseFile() error
}

type rocksdbFile struct {
  rdb *RocksDB
}

func (f *rocksdbFile) AppendFile(contents []byte) error {
  return C.DBEnvAppendFile(f.rdb, contents)
}
```

(This example is missing a lot of specifics, of course.)

That way you only need to add one method, OpenFile, to the engine interface. Unless there's a reason you didn't take this approach to begin with?

**Comment by bdarnell** [comment]
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/syncing_write.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBcKmTm9B1DnzK9Jg0U:-LBcKmTm9B1DnzK9Jg0V:b-o8673d) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd09093131a290b0d1f46618b8b18a4dec69be8/pkg/storage/syncing_write.go#L110)):*
> ```Go
> 		}
> 		if err == nil && sync {
> 			err = f.Sync()
> ```

This call to `Sync` was important for smoothing out performance in bulk IO tests. We need to keep that around. 

@dt What's the best way to test that we don't introduce performance regressions here?

---

*[pkg/storage/engine/engine.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBcJS-yB1nl1C41RKy3:-LBcJS-yB1nl1C41RKy4:bwc8iib) ([raw file](https://github.com/cockroachdb/cockroach/blob/9df1c349f2aaed78156e620b4fd4f9fc3d603416/pkg/storage/engine/engine.go#L36)):*
> ```Go
> // #include <stdlib.h>
> // #include <libroach.h>
> import "C"
> ```

We try not to duplicate all these headers in so many places. Also, `C` types are sometimes surprising in public interfaces. I think it would be best to define a type in rocksdb.go where the C stuff is already imported (especially with Nikhil's suggestion to put append and close on a new type). 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by windchan7** [comment]
@mberhault , Hey Marc, do you mind taking a look at this PR and let me know if I'm on the right track?

File: c-deps/libroach/engine.cc
Comment (mberhault) [inline c-deps/libroach/engine.cc]: make sure to run `make c-deps-fmt`, or `clang-format` directly, or add `clang-format` to your editor save hook for `.cc` and `.h` files.

File: c-deps/libroach/engine.cc
Comment (mberhault) [inline c-deps/libroach/engine.cc]: nit: we tend to use the full word `status` for clarity.

File: c-deps/libroach/engine.cc
Comment (mberhault) [inline c-deps/libroach/engine.cc]: for all the `if (!s.ok()) { return ToDBStatus(s); } return kSuccess;` calls, you can just return `ToDBStatus(s)`.

**Review by mberhault — COMMENTED** [review]
Overall looks good. I think a small benchmark comparing regular vs through-libroach writes would be good (one using an in-mem env, and one using actual disk).

**Comment by dt** [comment]
Review status: 3 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/storage/syncing_write.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBcKmTm9B1DnzK9Jg0U:-LCQ1FZW1tClebRSotof:b-dklvsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd09093131a290b0d1f46618b8b18a4dec69be8/pkg/storage/syncing_write.go#L110)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This call to `Sync` was important for smoothing out performance in bulk IO tests. We need to keep that around. 

@dt What's the best way to test that we don't introduce performance regressions here?
</blockquote></details>

looks like it was re-added back so I'm :thumbsup: here, but to answer w.r.t. testing:
the effect it had was harder to spot in isolated benchmarks, but was more apparent when running production tests, particularly on clouds the had less than ideal disk IO. When not syncing incrementally, we'd see more frequent slow liveness heartbeats, sometimes leading to cluster instability. TL;DR: If we were worried about ti, I'd run >1gb RESTORE (or IMPORT) on a cluster running with iffy disks and watch the latency of some minimal workload or for liveness errors. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by bdarnell** [comment]
Reviewed 16 of 16 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[c-deps/libroach/engine.h, line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8Fa22OV032t_VpLO:-LCR8Fa22OV032t_VpLP:bnqrs28) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.h#L46)):*
> ```objc
>   virtual DBStatus EnvWriteFile(DBSlice path, DBSlice contents) = 0;
>   virtual DBStatus EnvOpenFile(DBSlice path, rocksdb::WritableFile** file) = 0;
>   virtual DBStatus EnvAppendFile(rocksdb::WritableFile** file, DBSlice contents) = 0;
> ```

`EnvOpenFile` needs a double-pointer, but I think the rest of these should just be a single pointer. 

---

*[c-deps/libroach/engine.cc, line 233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8TDJ-9u-oVaWeWhi:-LCR8TDJ-9u-oVaWeWhj:beom0yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L233)):*
> ```cpp
> 
>   // Recursively create the directory of the given file.
>   std::string delimiter = "/";
> ```

Not true on windows. Does rocksdb have a function we could use that does the right thing here (either to parse a string into directories or take a string and create multiple directories at once)?

---

*[pkg/storage/client_raft_test.go, line 1142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAZdn18dC_-zvvW6F:-LCRAZdn18dC_-zvvW6G:b-k19ihl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_raft_test.go#L1142)):*
> ```Go
> 	defer mtc.Stop()
> 
> 	for i := 0; i < 3; i++ {
> ```

Why is this added to this test?

---

*[pkg/storage/client_replica_gc_test.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAed89IEsZxNl_zrb:-LCRAed89IEsZxNl_zrc:b-1kzc8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_replica_gc_test.go#L75)):*
> ```Go
> 	defer mtc.Stop()
> 
> 	for i := 0; i < numStores; i++ {
> ```

And here?

---

*[pkg/storage/replica_sideload_test.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRBNywAnHNEca7xkDQ:-LCRBNywAnHNEca7xkDR:b-qbspyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/replica_sideload_test.go#L111)):*
> ```Go
> 	st := cluster.MakeTestingClusterSettings()
> 
> 	cache := engine.NewRocksDBCache(1 << 20)
> ```

It would be nice to factor out some of the new repetition in this file.

---

*[pkg/storage/engine/rocksdb.go, line 2623 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAxxG2--6q8xje1BH:-LCRAxxG2--6q8xje1BI:b-bufo6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/engine/rocksdb.go#L2623)):*
> ```Go
> func (r *RocksDB) OpenFile(filename string) (DBFile, error) {
> 	var file C.DBWritableFile
> 	/*
> ```

Uncomment or remove this before merging. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by windchan7** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8TDJ-9u-oVaWeWhi:-LCTygv46iIIUXIQLK2Z:b-yvrq2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L233)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Not true on windows. Does rocksdb have a function we could use that does the right thing here (either to parse a string into directories or take a string and create multiple directories at once)?
</blockquote></details>

I didn't find any method on the RocksDB side that does the job. I also asked Marc and we agreed to write the recursive directory creation method on our own. 

Do you have any suggestions on how to work around it, Ben? I feel like adding support for both Windows and Linux could cause a mess.

---

*[pkg/storage/client_raft_test.go, line 1142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAZdn18dC_-zvvW6F:-LCTyXYdA_AdK4FGoShu:bad6l0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_raft_test.go#L1142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is this added to this test?
</blockquote></details>

It's because if not, the test will use InMem RocksDB, which does not have access to the disk.

---

*[pkg/storage/client_replica_gc_test.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAed89IEsZxNl_zrb:-LCTycCz8IaLM8zZfKWv:bi1q2rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_replica_gc_test.go#L75)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

And here?
</blockquote></details>

Same reason as above.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by windchan7** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8TDJ-9u-oVaWeWhi:-LCUCMwJDdOl_GpoJDPc:bfew1b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L233)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I saw that you had some commented-out code to create the directory from Go. That would be the simplest thing; is there some reason we can't use it?
</blockquote></details>

It's mainly because the `os.Mkdir` from go only works for on disk RocksDB. For inMem RocksDB, it won't notice there's a directory on disk already. Instead, it will throw a `directory does not exist` error, or something like that.

---

*[pkg/storage/client_raft_test.go, line 1142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAZdn18dC_-zvvW6F:-LCUByFGApH8HFYdPWuS:b-j4ke5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_raft_test.go#L1142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why does it need access to the disk now but it didn't before? If we can't mock everything out in memory, I'd rather change multiTestContext so that all the tests use on-disk rocksdb instead of having a few tests work differently. 
</blockquote></details>

Before the changes, writeFileSyncing always writes to disk regardless its inMem or on disk RocksDB. However, since we want everything to be written to RocksDB's env, inMem env won't be able to write to disk so the checks in the tests will fail if we still use inMem RocksDB.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8TDJ-9u-oVaWeWhi:-LCUV5b1Di4XJ9yTIHam:bg8uh23) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L233)):*
<details><summary><i>Previously, windchan7 (Victor Chen) wrote…</i></summary><blockquote>

It's mainly because the `os.Mkdir` from go only works for on disk RocksDB. For inMem RocksDB, it won't notice there's a directory on disk already. Instead, it will throw a `directory does not exist` error, or something like that.
</blockquote></details>

What do you mean it won't notice there's a directory on disk already? Are the Go and C++ sides looking at the same path?

---

*[pkg/storage/client_raft_test.go, line 1142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAZdn18dC_-zvvW6F:-LCUVHnaChEBLzlzvVNx:b-ukg14a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_raft_test.go#L1142)):*
<details><summary><i>Previously, windchan7 (Victor Chen) wrote…</i></summary><blockquote>

Before the changes, writeFileSyncing always writes to disk regardless its inMem or on disk RocksDB. However, since we want everything to be written to RocksDB's env, inMem env won't be able to write to disk so the checks in the tests will fail if we still use inMem RocksDB.
</blockquote></details>

Why don't other tests in this file need this too? I don't like having a couple of tests with a 20-line copy-pasted block of boilerplate. We should either support these methods on InMem RocksDB engines or move multiTestContext to use on-disk rocksdb in all cases. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by windchan7** [comment]
Commenting out `TestReplicaLazyLoad` for now because of this issue: 
https://github.com/cockroachdb/cockroach/issues/25430

**Comment by windchan7** [comment]
Review status: 2 of 18 files reviewed at latest revision, 11 unresolved discussions.

---

*[c-deps/libroach/engine.h, line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8Fa22OV032t_VpLO:-LCVe5FO4keRb5NktPfZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.h#L46)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`EnvOpenFile` needs a double-pointer, but I think the rest of these should just be a single pointer. 
</blockquote></details>

Done.

---

*[c-deps/libroach/engine.cc, line 228 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBaV94l3krDtKxuve-z-r2-228:-LCVe64w56FWihA5pU1P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L228)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

nit: we tend to use the full word `status` for clarity.
</blockquote></details>

Done.

---

*[c-deps/libroach/engine.cc, line 233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCR8TDJ-9u-oVaWeWhi:-LCVe6zkCtauGVQT6ZU6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L233)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What do you mean it won't notice there's a directory on disk already? Are the Go and C++ sides looking at the same path?
</blockquote></details>

Done.

---

*[c-deps/libroach/engine.cc, line 238 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBaV94l3krDtKxuve-z-r2-238:-LCVe89T66-dMLqCihnM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L238)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

make sure to run `make c-deps-fmt`, or `clang-format` directly, or add `clang-format` to your editor save hook for `.cc` and `.h` files.
</blockquote></details>

Done.

---

*[c-deps/libroach/engine.cc, line 269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBaV94l3krDtKxuve-z-r2-269:-LCVe9vlChhRG7XeBPZK:b-he8ayz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/c-deps/libroach/engine.cc#L269)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

for all the `if (!s.ok()) { return ToDBStatus(s); } return kSuccess;` calls, you can just return `ToDBStatus(s)`.
</blockquote></details>

Changed at necessary places.

---

*[pkg/storage/client_raft_test.go, line 1142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAZdn18dC_-zvvW6F:-LCVezkM3uRUmVrIfeDX:bxjubb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/client_raft_test.go#L1142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why don't other tests in this file need this too? I don't like having a couple of tests with a 20-line copy-pasted block of boilerplate. We should either support these methods on InMem RocksDB engines or move multiTestContext to use on-disk rocksdb in all cases. 
</blockquote></details>

I made all multiTestContext using on disk engines. Thanks for the suggestion.

---

*[pkg/storage/replica_sideload_test.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRBNywAnHNEca7xkDQ:-LCVeKFq6k8txP7YQg74:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/replica_sideload_test.go#L111)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It would be nice to factor out some of the new repetition in this file.
</blockquote></details>

Done.

---

*[pkg/storage/engine/engine.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LBcJS-yB1nl1C41RKy3:-LCVexKx1thbvG-VqFNW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9df1c349f2aaed78156e620b4fd4f9fc3d603416/pkg/storage/engine/engine.go#L36)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We try not to duplicate all these headers in so many places. Also, `C` types are sometimes surprising in public interfaces. I think it would be best to define a type in rocksdb.go where the C stuff is already imported (especially with Nikhil's suggestion to put append and close on a new type). 
</blockquote></details>

Done.

---

*[pkg/storage/engine/rocksdb.go, line 2623 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25281#-LCRAxxG2--6q8xje1BH:-LCVeyPTBN2BFnRDti2R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0378351bf6280a6caa11012991510dba7cc53f3/pkg/storage/engine/rocksdb.go#L2623)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Uncomment or remove this before merging. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25281)*

**Comment by windchan7** [comment]
@mberhault Is it ready to go? I'm holding off merging it until the flaky test get resolved. But feel free to comment on the rest of it.

File: c-deps/libroach/engine.cc
Comment (mberhault) [inline c-deps/libroach/engine.cc]: This can be removed now.

File: c-deps/libroach/engine.cc
Comment (mberhault) [inline c-deps/libroach/engine.cc]: The file object still needs to be deleted on error from `Close`, or we'll be leaking memory.

**Comment by windchan7** [comment]
@mberhault Now the flaky test is resolved.

**Review by mberhault — APPROVED** [review]
LGTM

**Pull request merged**
